apacheconfig/wloader.py

```diff
@@ -31,19 +31,135 @@ def dump(self):
     def typestring(self):
         """Returns object typestring as defined by the apacheconfig parser."""
 
-    @abc.abstractproperty
-    def whitespace(self):
-        """Returns preceding or trailing whitespace for this node as a string.
+
+class ListNode(AbstractASTNode):
+    """Creates object representing an ordered list of LeafASTNodes.
+
+    Each BlockNode contains a ListNode, and every configuration file's
+    root should be a ListNode. To construct from a raw string, use the
+    `parse` factory function. The default __init__ constructor expects data
+    from the internal apacheconfig parser.
+
+    Args:
+        raw (list): Raw data returned from ``apacheconfig.parser``.
+    """
+    def __init__(self, raw, parser):
+        self._type = raw[0]
+        self._contents = []
+        self._trailing_whitespace = ""
+        self._parser = parser
+        for elem in raw[1:]:
+            if isinstance(elem, str) and elem.isspace():
+                self._trailing_whitespace = elem
+            elif elem[0] == "block":
+                self._contents.append(BlockNode(elem, parser))
+            else:
+                self._contents.append(LeafASTNode(elem))
+
+    @classmethod
+    def parse(cls, raw_str, parser):
+        """Factory for :class:`apacheconfig.ListASTNode` from a config string.
+
+        Args:
+            raw_str (str): Config string to parse.
+            parser (:class:`apacheconfig.ApacheConfigParser`): parser object
+                to use.
+        Returns:
+            :class:`apacheconfig.ListASTNode` containing data parsed from
+            ``raw_str``.
         """
+        raw = parser.parse(raw_str)
+        return cls(raw, parser)
 
-    @abc.abstractmethod
-    @whitespace.setter
-    def whitespace(self, value):
-        """Set preceding or trailing whitespace for this node.
+    def add(self, index, raw_str):
+        """Parses and adds child element at given index.
+
+        Parses given string into an ASTNode object, then adds to list at
+        specified index.
 
         Args:
-            value (str): value to set whitespace to.
+            index (int): index of list at which to insert the node.
```

Comment:
I left them as `IndexError` since it describes a class of errors that's different from `ApacheConfigError`, which seems to be reserved for strange parsing bugs or malformed configuration files.